File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-1.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int l= heights.length;
        int right[]=new int[l];
        int left[]=new int[l];

        Stack<Integer>st=new Stack<>();

        // right smaller element
        for(int i=l-1;i>=0;i--){
            while(!st.isEmpty() && heights[st.peek()]>= heights[i]){
                st.pop();
            }

            if(st.isEmpty()){
                right[i]=l;
            }
            else{
                right[i]=st.peek();
            }

            st.push(i);
        }

        // left smaller element
        st=new Stack<>();

        for(int i=0;i<l;i++){
            while(!st.isEmpty() && heights[st.peek()]>= heights[i]){
                st.pop();
            }

            if(st.isEmpty()){
                left[i]=-1;
            }
            else{
                left[i]=st.peek();
            }

            st.push(i);
        }

        int ans=0;

        for(int i=0;i<l;i++){
            int w=right[i]-left[i]-1;
            ans=Math.max(ans,heights[i]*w);
        }

        return ans;
    }
}

File: Data Structures & Algorithms/spiral-matrix/submission-1.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& m) {
        
        int n1 = m.size() , n2 = m[0].size();
        int l = 0 , r = n2-1 , t = 0 , d = n1-1;
        vector<int>ans;
        while(r>=l && d>=t)
        {
            // top
            for(int i=l;i<=r && d>=t;i++) ans.push_back(m[t][i]);
            t++;
            //right
            for(int i=t;i<=d && r>=l;i++) ans.push_back(m[i][r]);
            r--;
            //down
            for(int i=r;i>=l && d>=t;i--) ans.push_back(m[d][i]);
            d--;
            //left
            for(int i=d;i>=t && r>=l;i--) ans.push_back(m[i][l]);
            l++;
        }
        return ans;
    }
};
